examples/alignment.dx

```diff
@@ -0,0 +1,381 @@
+
+Hertz = Float
+Mel = Float
+def mel (f:Hertz) : Mel = 1125.0 * log (1.0 + (f / 700.0))
+def imel (f:Mel) : Hertz = 700.0 * (exp (f / 1125.0) - 1.0)
+
+' However it is now so easy to move between frequency scales. To do this we will need
+' a bit of Dex machinery.
+
+' ### Linear Scales
+
+' Let us define a scaled range as a linear mapping between ordinals and a range. 
+
+LinearScale = {start:Float & end:Float}
+data ScaledRange a = AsScaledRange scale:LinearScale
+
+' Given a scale, we can easily map from an index to a point, get the step, get the full domain,
+  and take the binned inverse to get the nearest index.
+
+def step (AsScaledRange {start, end}:ScaledRange a) : Float =
+    (1.0 / (IToF ((size a) - 1) )) * (end - start)
+def scaled (AsScaledRange {start, end}:ScaledRange a) (x:a): Float =
+    start + ((IToF (ordinal x)) / (IToF ((size a) - 1))) * (end - start)
+def domain (scale:ScaledRange a) : a => Float =
+    for i. scaled scale i
+def bin (scale:ScaledRange a) (x:Float) : a =
+    (AsScaledRange {start, end}) = scale
+    (FToI (floor ((x - start) / (step scale))))@_
+
+' ## Mapping Between Scales
+
+' We can then define a two linear ranges. One in Hertz (corresponding to the FFT) and one in Mel Space.
+
+MelBins = Fin 28
+
+hscale : ScaledRange Positive = AsScaledRange {start=0.0, end=samplerate / 2.0}
+melscale : ScaledRange MelBins = AsScaledRange {start=mel 0.0, end=mel (samplerate / 2.0)}
+domain melscale
+
+mel_to_hscale = for i: MelBins.
+        mel = scaled melscale i
+        bin hscale (imel mel)
+
+
+' ## Triangle Filter Bank
+
+' In order to apply this mapping we need a way to handle the non-linear disconnect. To do this
+  We build a triangle windowing function.
+
+:t fft_frames
+
+def triangle_window (mpt:m) (mlower:m) (mupper:m) (scale: ScaledRange m) : m => Float =
+    pt = IToF (ordinal mpt)
+    lower = IToF (ordinal mlower)
+    upper = IToF (ordinal mupper)
+    for i'.
+        i = IToF (ordinal i')
+        if (i >= lower) && (i <= pt) 
+           then (i - lower) / (pt - lower)
+           else if (i <= upper) && (i >= pt) 
+             then (upper - i) / (upper - pt)
+             else 0.0
+
+' The 25'th Mel bin is going to take in the mass from this range of hertz's bins.
+
+:html showPlot $ xyPlot (domain hscale) (triangle_window mel_to_hscale.(25@_) mel_to_hscale.(24@_) mel_to_hscale.(26@_) hscale)
+
+' Whereas the 19'th Mel bin is going to take in the mass from this range of hertz's bins.
+
+:html showPlot $ xyPlot (domain hscale) (triangle_window mel_to_hscale.(19@_) mel_to_hscale.(18@_) mel_to_hscale.(20@_) hscale)
+
+
+' Here is the full matrix.
+
+start = 1@MelBins
+end = 27@MelBins
+MelBinsIn = Fin 26
+H : MelBinsIn => Positive => Float = unsafeCastTable MelBinsIn $ for m' : (start..<end).
+      m = %inject m'
+      m1 = ((ordinal m) - 1)@MelBins
+      m2 = ((ordinal m) + 1)@MelBins
+      triangle_window mel_to_hscale.m mel_to_hscale.m1 mel_to_hscale.m2 hscale
```

Comment:
Couldn't this be expressed as a map over `window mel_to_hscale`?

Reply:
Almost! The problem here is that the windows are of differing width size.  

See diagram -> https://stackoverflow.com/questions/40197060/librosa-mel-filter-bank-decreasing-triangles

However, there must be a way to do better. Windowing and padding seem like the key.

Reply:
Sure, they "triangle windows" are of varying size, but the windowing on `mel_to_hscale` is regular! You always take `m`, `m - 1` and `m + 1`!

Reply:
I'm thinking of something along the lines of `map (triangle_window hscale) $ window 1 mel_to_hscale`